### PR TITLE
Feature/PandaIKFast

### PR DESCRIPTION
**Pull request by Joobz**
Created folder _pandarobot_ (in the style of _cablerobot_), which has the generated ikfast cpp file for the Panda robot, its header file and a wrapper so it can be easily used. There are also unittests for two wrapped functions (FK and IK).
The wrapper is not yet adapted to the GT types.

### CODE REVIEW

File: gtdynamics/pandarobot/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# add cablerobot subfolders to gtdynamics' SOURCE_SUBDIRS list
+list(APPEND SOURCE_SUBDIRS pandarobot/ikfast)
+set(SOURCE_SUBDIRS ${SOURCE_SUBDIRS} PARENT_SCOPE)
+
+# add wrapper interface file
+# ist(APPEND ADDITIONAL_INTERFACE_FILES ${CMAKE_CURRENT_SOURCE_DIR}/cablerobot.i)
```
Comment (dellaert) [inline gtdynamics/pandarobot/CMakeLists.txt:6]: Change to panda robot.i

File: gtdynamics/pandarobot/tests/testPandaIKfast.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+#include <iostream>
+#include <vector>
+
+using namespace gtsam;
+using gtsam::assert_equal;
+
+TEST(IkFastWrapper, Forward) {
+    // Create ikfastwrapper object
+    panda::IkFastWrapper pandarobot = panda::IkFastWrapper();
```
Comment (dellaert) [inline gtdynamics/pandarobot/tests/testPandaIKfast.cpp:17]: rename class

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+using IKFAST_NAMESPACE::IkReal;
+
+
+namespace panda { 
+    using ikvector = std::vector<IkReal>;
+
+    class IkFastWrapper { 
+        public:
+            int num_joints, num_free_params; //should be static
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:81]: Constexpr size_r static, kNumJoints (check Google c++ style) = 7

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+#include <stdio.h>
+#include <iostream>
+#include <stdlib.h>
+#include <vector>
+
+using IKFAST_NAMESPACE::IkReal;
+
+
+namespace panda { 
+    using ikvector = std::vector<IkReal>;
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:77]: JointAngles

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            num_free_params = IKFAST_NAMESPACE::GetNumFreeParameters();
+    } 
+
+    /**
+     * @brief Forward IkFastWrapper on Panda robot using IKFast.
+     * 
+     * @param joint_config 
+     * @return std::vector<IkReal> 
+     */
+    ikvector IkFastWrapper::forward(ikvector joint_config) {
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:98]: Static, const &

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        // Probably should be an assert
+        if( joint_config.size() != num_joints ) {
+            std::cout << "\nError: (forward IkFastWrapper) expects vector of "<< num_joints << " values describing joint angles (in radians).\n\n";
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:102]: Format with Google style

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h (hunk 4 above)
Comment (dellaert) [reply gtdynamics/pandarobot/ikfast/ikfastwrapper.h:77]: Better yet, use a gtsam::Vector7

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            return ikvector(0);
+        }
+
+        // This will be changed once completely integrated
+        // Put input joint values into array, could be better
+        IkReal joints[num_joints];
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:108]: Const

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        for (int i=0; i<num_joints; i++)
+        {
+            joints[i] = joint_config[i];
+        }
+
+        IkReal eerot[9],eetrans[3];
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:114]: Rename to orientation, position, and add comment

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h (hunk 5 above)
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:98]: Return Pose3

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        IKFAST_NAMESPACE::ComputeFk(joints, eetrans, eerot);
+
+
+        // This will be changed once completely integrated
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:118]:
```
Rot3 bRe(…);
Point3 bte (…);
Return Pose3(bRe,bte);
```

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        return ee_pose;
+    }
+    
+    /**
+     * @brief Inverse IkFastWrapper on Panda robot using IKFast.
+     * 
+     * @param ee_pose 
+     * @param vfree
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:134]: Explain

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    }
+    
+    /**
+     * @brief Inverse IkFastWrapper on Panda robot using IKFast.
+     * 
+     * @param ee_pose 
+     * @param vfree 
+     * @return std::vector<std::vector<IkReal>>
+     */
+    std::vector<ikvector> IkFastWrapper::inverse(ikvector ee_pose, ikvector vfree) {
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:137]: const Pose3& bTe, double theta7

File: gtdynamics/pandarobot/ikfast/ikfastwrapper.h [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        bool bSuccess = IKFAST_NAMESPACE::ComputeIk(eetrans, eerot, vfree.size() > 0 ? &vfree[0] : NULL, solutions);
+
+        if( !bSuccess ) {
+            fprintf(stderr,"Error: (inverse IkFastWrapper) failed to get ik solution\n");
+            return std::vector<ikvector>();
+        }
+
+        int num_sols = (int)solutions.GetNumSolutions();
+
+        // This will be changed once completely integrated
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/ikfastwrapper.h:170]: Just return the vector of Vector7, empty if there is only singularities or outside workspace

**Comment by dellaert** [comment]
Also check naming conventions of files and definitely format the non-generated files with Google style

**Review by dellaert — APPROVED** [review]

File: gtdynamics/pandarobot/ikfast/PandaIKFast.h [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+//----------------------------------------------------------------------------//
+
+#include <gtsam/base/Vector.h>
+#include <gtsam/geometry/Pose3.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <iostream>
+#include <vector>
+
```
[the remark this answers is not in the document]
Comment (Joobz) [reply gtdynamics/pandarobot/ikfast/PandaIKFast.h:26]: Okay, got it! Going to push a new commit then.

**Comment by Joobz** [comment]
Just solved the comments, going to request a review  (@dellaert ) again just in case. I didn't merge yet to leave all the pandaIKFast related things in this PR, since there are nearly no changes from the previous commit.

File: gtdynamics/pandarobot/ikfast/PandaIKFast.cpp [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+#include <stdlib.h>
+
+#include <iostream>
+#include <vector>
+
+namespace gtdynamics {
+
+PandaIKFast::PandaIKFast() {}
+
+gtsam::Pose3 PandaIKFast::forward(const gtsam::Vector7& joint_values) {
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/PandaIKFast.cpp:33]: In cpp you should use `using` as much as possible :-) at top of file, one line per gtsam class, e.g.,

```
using gtsam::Pose3;
```

still no blanket namespace using….

File: gtdynamics/pandarobot/ikfast/PandaIKFast.cpp [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  // The inputs (except "solutions") have to be arrays
+  bool success = panda_internal::ComputeIk(bTe.translation().data(),
+                                             bRe.data(), &theta7, solutions);
+
+  if (!success) {
+    fprintf(stderr, "Error: (inverse PandaIKFast) failed to get ik solution\n");
+    return std::vector<gtsam::Vector7>();
+  }
+
+  unsigned int num_sols = (int)solutions.GetNumSolutions();
```
Comment (dellaert) [inline gtdynamics/pandarobot/ikfast/PandaIKFast.cpp:66]: Cast does not match type

**Review by dellaert — APPROVED** [review]
Approved. After addressing final comment, just merge :-)

File: gtdynamics/pandarobot/ikfast/PandaIKFast.cpp (hunk 14 above)
Comment (Joobz) [reply gtdynamics/pandarobot/ikfast/PandaIKFast.cpp:33]: Ooh, finally got it. I did not understand the reason earlier, just thought that it was arbitrary. Now it makes sense :-)

**Comment by dellaert** [comment]
Woo hoo, congrats on your first PR merge!

**Pull request merged**